modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeTableViewTest.java

```diff
@@ -374,83 +374,133 @@ private String debug() {
+
+        TreeItem<String> treeRoot = new TreeItem<String>("root");
+        treeRoot.setExpanded(true);
+        treeTableView.setRoot(treeRoot);
+
+        final int FIRST_LEVEL_COUNT = 8;
+        for (int i = 0; i < FIRST_LEVEL_COUNT; i++) {
+            TreeItem<String> ti = new TreeItem<>( "" + i);
+            ti.setExpanded(true);
+            treeRoot.getChildren().add(ti);
+
+            for (int j = 0; j < FIRST_LEVEL_COUNT - 1; j++) {
+                TreeItem<String> tj = new TreeItem<>("" + i + j);
+                tj.setExpanded(true);
+                ti.getChildren().add(tj);
+
+                for (int k = 0; k < FIRST_LEVEL_COUNT - 2; k++) {
+                    TreeItem<String> tk = new TreeItem<>("" + i + j + k);
+                    tk.setExpanded(true);
+                    tj.getChildren().add(tk);
+
+                    for (int l = 0; l < FIRST_LEVEL_COUNT - 3; l++) {
+                        TreeItem<String> tl = new TreeItem<>("" + i + j + k + l);
+                        tl.setExpanded(true);
+                        tk.getChildren().add(tl);
+
+                        for (int m = 0; m < FIRST_LEVEL_COUNT - 4; m++) {
+                            TreeItem<String> tm = new TreeItem<>("" + i + j + k + l + m);
+                            tl.getChildren().add(tm);
+                        }
+                    }
+                }
+            }
+        }
 
         sm.setSelectionMode(SelectionMode.MULTIPLE);
-        sm.selectIndices(2, 5, 8, 10);
+        int indices[] = new int[] {1, 400, 800, 1200, 1600, 2000, 2400, 2800, 3200, 3600, 4000, 4400, 4800, 5200, 5600, 6000, 6400};
+        sm.selectIndices(1, 400, 800, 1200, 1600, 2000, 2400, 2800, 3200, 3600, 4000, 4400, 4800, 5200, 5600, 6000, 6400);
 
         // Sanity checks
-        assertEquals(4, sm.getSelectedIndices().size());
-        assertEquals(10, sm.getSelectedIndex());
-        assertEquals(treeTableView.getTreeItem(10), sm.getSelectedItem());
+        assertEquals(indices.length, sm.getSelectedIndices().size());
+        assertEquals(indices.length, sm.getSelectedItems().size());
+        assertEquals(indices.length, sm.getSelectedCells().size());
+        assertEquals(indices[indices.length - 1], sm.getSelectedIndex());
+        assertEquals(treeTableView.getTreeItem(indices[indices.length - 1]), sm.getSelectedItem());
 
-        selectedItem = (TreeItem<String>) sm.getSelectedItem();
-        selectedCellsBeforePermutation = new ArrayList<>(sm.getSelectedCells());
+        selectedItemBefore = (TreeItem<String>) sm.getSelectedItem();
+        selectedItemsBefore = new ArrayList<>(sm.getSelectedItems());
+        selectedIndicesBefore = new ArrayList<>(sm.getSelectedIndices());
+        selectedCellsBefore = new ArrayList<>(sm.getSelectedCells());
 
         sm.selectedIndexProperty().addListener(ov -> {
             countSelectedIndexChangeEvent++;
+            assertEquals(selectedItemBefore, treeTableView.getTreeItem(sm.getSelectedIndex()));
```

Comment:
If this assertion ever fails, I don't think it will cause a test failure, since the event handling code will swallow the exception. The same is true of the other listeners. I don't know if it would be possible to use the UncaughtExceptionHandler as is done in other tests -- see [JDK-8244531](https://bugs.openjdk.java.net/browse/JDK-8244531) -- but that might be worth exploring. Another possibility is to wrap all the listeners in a try/catch and keep a list of `Throwable`s that are caught.

Reply:
Hi Kevin, I tested these listeners by adding assert that always fail. It shows the failures correctly. Also in this file there are other tests which assert inside a listener. I confirmed that those tests also show failures correctly. I could not find out what makes it work though.

Reply:
I can confirm that the test fails if an assertion hits. The reason is that this ChangeListener throws any exception to the code that causes the change, so it will propagate to the test thread.